File: libraries/dotnet/logs/src/LogSampler.cs
// This file contains log sampler logic for logs src.
namespace Otel.Logs;

internal sealed class LogSampler
{
    private readonly Dictionary<string, LogBatch> batchMap = new(StringComparer.Ordinal);
    private readonly double probabilisticSamplingRate = LogsConfiguration.SamplingRate();
    private readonly Logger logger;
    private readonly object syncRoot = new();

    // Logs sampler.
    public LogSampler(Logger logger)
    {
        this.logger = logger;
    }

    // Adds log.
    public void AddLog(LogEntry logEntry)
    {
        lock (syncRoot)
        {
            if (probabilisticSamplingRate == 0)
            {
                logger.ProcessLog(logEntry);
                return;
            }

            if (!LogsConfiguration.HasValue(logEntry.InvocationId) || logEntry.InvocationId == "unknown")
            {
                logger.ProcessLog(logEntry);
                return;
            }

            if (batchMap.TryGetValue(logEntry.InvocationId, out var existingBatch))
            {
                existingBatch.Logs.Add(logEntry);
                return;
            }

            FlushOneBatch();
            batchMap[logEntry.InvocationId] = new LogBatch(logEntry);
        }
    }

    // Flushes one batch.
    public void FlushOneBatch()
    {
        lock (syncRoot)
        {
            foreach (var (invocationId, batch) in batchMap.ToArray())
            {
                var hasError = batch.Logs.Any(log => log.Level == LogLevel.Error);
                if (hasError || ShouldSample())
                {
                    foreach (var log in batch.Logs)
                    {
                        logger.ProcessLog(log);
                    }
                }

                batchMap.Remove(invocationId);
            }
        }
    }

    // Decides whether to sample.
    private bool ShouldSample()
    {
        var threshold = Math.Clamp(probabilisticSamplingRate, 0, 100) / 100;
        return Random.Shared.NextDouble() <= threshold;
    }
}

File: libraries/java/logs/src/main/java/otel/logs/LogSampler.java
// This file contains log sampler logic for OTel logs.
package otel.logs;

import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

final class LogSampler {
  private final Map<String, LogBatch> batchMap = new LinkedHashMap<>();
  private final double probabilisticSamplingRate = LogsConfiguration.samplingRate();
  private final Logger logger;

  LogSampler(Logger logger) {
    this.logger = logger;
  }

  synchronized void addLog(LogEntry logEntry) {
    if (probabilisticSamplingRate == 0) {
      logger.processLog(logEntry);
      return;
    }

    if (!LogsConfiguration.hasValue(logEntry.invocationId) || "unknown".equals(logEntry.invocationId)) {
      logger.processLog(logEntry);
      return;
    }

    LogBatch existingBatch = batchMap.get(logEntry.invocationId);
    if (existingBatch != null) {
      existingBatch.logs.add(logEntry);
      return;
    }

    flushOneBatch();
    batchMap.put(logEntry.invocationId, new LogBatch(logEntry));
  }

  synchronized void flushOneBatch() {
    if (batchMap.isEmpty()) return;

    Iterator<Map.Entry<String, LogBatch>> iterator = batchMap.entrySet().iterator();
    while (iterator.hasNext()) {
      Map.Entry<String, LogBatch> entry = iterator.next();
      LogBatch batch = entry.getValue();
      boolean hasError = batch.logs.stream().anyMatch(log -> log.level == LogLevel.ERROR);

      if (hasError || shouldSample()) {
        batch.logs.forEach(logger::processLog);
      }

      iterator.remove();
    }
  }

  // Decides whether to sample.
  private boolean shouldSample() {
    double threshold = Math.max(0, Math.min(100, probabilisticSamplingRate)) / 100.0;
    return Math.random() <= threshold;
  }
}
